luci2-ui-base/src/www/luci-ng/uci/uciValidation.js

```diff
@@ -0,0 +1,483 @@
+
+					if (self.message)
+						msgs.push(self.message.format.apply(self.message, arguments[i+1]));
+				}
+
+				self.message = msgs.join(', ');
+				return true;
+			},
+
+			'neg': function() {
+				return self.types['or'].apply(
+					this.replace(/^[ \t]*![ \t]*/, ''), arguments);
+			},
+
+			'list': function(subvalidator, subargs) {
+				if (typeof subvalidator != 'function')
+					return false;
+
+				var tokens = this.match(/[^ \t]+/g);
+				for (var i = 0; i < tokens.length; i++)
+					if (!subvalidator.apply(tokens[i], subargs))
+						return false;
+
+				return true;
+			},
+
+			'phonedigit': function() {
+				if (this.match(/^[0-9\*#!\.]+$/) != null)
+					return true;
+
+				self.message = gettext('Must be a valid phone number digit');
+				return false;
+			},
+
+			'string': function() {
+				return true;
+			}
+		};
+
+		this.compile(code);
+	}
+
+	Validation.prototype = {
+		message: undefined,
+		result: undefined,
+		_stack: null,
+
+		compile: function(code) {
+			this.code = code;
+			this._stack = this._compile(code);
+		},
+
+		_compile: function(code) {
+			var pos = 0;
+			var esc = false;
+			var depth = 0;
+			var stack = [];
+
+			code += ',';
+
+			for (var i = 0; i < code.length; i++) {
+				if (esc) {
+					esc = false;
+					continue;
+				}
+
+				switch (code.charCodeAt(i)) {
+					case 92: // '\'
+						esc = true;
+						break;
+
+					case 40: // '('
+					case 44: // ','
+						if (depth <= 0) {
+							if (pos < i) {
+								var label = code.substring(pos, i);
+								label = label.replace(/\\(.)/g, '$1');
+								label = label.replace(/^[ \t]+/g, '');
+								label = label.replace(/[ \t]+$/g, '');
+
+								if (label && !isNaN(label)) {
+									stack.push(parseFloat(label));
+								} else if (label.match(/^(['"]).*\1$/)) {
+									stack.push(label.replace(/^(['"])(.*)\1$/, '$2'));
+								} else if (typeof this.types[label] == 'function') {
+									stack.push(this.types[label]);
+									stack.push([]);
+								} else							{
+									throw new Error('Syntax error, unhandled token \''+label+'\'');
+								}
+							}
+							pos = i+1;
+						}
+						depth += (code.charCodeAt(i) == 40);
+						break;
+
+					case 41: // ')'
+						if (--depth <= 0) {
+							if (typeof stack[stack.length-2] != 'function')
+								throw new Error('Syntax error, argument list follows non-function');
+
+							stack[stack.length-1] = this._compile(code.substring(pos, i));
+
+							pos = i+1;
+						}
+						break;
+				}
+			}
+
+			return stack;
+		},
+
+		test: function(value) {
+			this.message=null;
+
+			if (this._stack &&
+				!this._stack[0].apply(value, this._stack[1])) {
+				this.result = false;
+				this.message = this.message || gettext('Validation error');
+				return this.message;
+			}
+
+			this.result = true;
+
+			return undefined;
+		},
+
+		getValidator: function(modelCtrl) {
+			var self = this;
+
+			return function(modelValue, viewValue) {
+				modelCtrl.uciValidator = '';
```

Comment:
The double meaning of `uciValidator` confused me here. Within this module it appears to be both used for a uci validator instance and for the resulting error message. Maybe we should rename the message property to `uciValidationResult` or `uciValidationMessage` ?